### ISSUE TITLE
Information on build process

### ISSUE THREAD

**Issue by paulmaunders** [marcinbojko/vagrant-boxes#1]
Hi,

Thank you very much for providing these boxes. You RockyLinux 9 vagrant box is the first I have found which works on my M2 Mac (arm64).

I was wondering if you have any documentation or source code available for the build process? I see you have a Hyper-V packer project, but I couldn't see any code for how you are building the parallels boxes.  

In particular, I am interested to know how you got the parallels tools to install as they seem to fail with RHEL 9 using other packer projects.

Thanks,

Paul Maunders

**Comment by marcinbojko** [comment]
Hi @paulmaunders 
I have a separate repo for parallels and VirtualBox, but they share the same /extras folder.
Unfortunately, they are too buggy to make them public but if you require some parallels-related stuff I can share these.

**Comment by marcinbojko** [comment]
https://github.com/marcinbojko/hv-packer/blob/master/extra/files/gen2-linux/parallels.sh

**Comment by paulmaunders** [comment]
Thanks @marcinbojko do you know which version of parallels tools you are using? When I tried to build a box with Bento it failed https://forum.parallels.com/threads/parallels-tools-do-not-install-on-rhel9.357546/page-2#post-910871 

```
parallels-iso: Started installation of Parallels Guest Tools version '18.1.0.53311'
parallels-iso: modprobe: FATAL: Module prl_tg not found in directory /lib/modules/5.14.0-70.30.1.el9_0.aarch64
```
This was using `prl-tools-lin-arm.iso` and Parallels Tools Version: `18.1.0.53311`

**Comment by marcinbojko** [comment]
Hi @paulmaunders  - Version 18.1.0 (53311)
But there is a flaw in parallels SDK - packer 1.8.4 doesn't see it, so you have to use some workarounds with symbolic links

**Comment by marcinbojko** [comment]
@paulmaunders 
```
    parallels-iso.autogenerated_1: Started installation of Parallels Guest Tools version '18.1.0.53311'
    parallels-iso.autogenerated_1: modprobe: FATAL: Module prl_tg not found in directory /lib/modules/5.14.0-70.13.1.el9.aarch64
    parallels-iso.autogenerated_1:
    parallels-iso.autogenerated_1: Thu Nov 24 13:39:45 CET 2022
    parallels-iso.autogenerated_1: Start installation or upgrade of Guest Tools
    parallels-iso.autogenerated_1: Skipping BTF generation for /usr/lib/parallels-tools/kmods/prl_tg/Toolgate/Guest/Linux/prl_tg/prl_tg.ko due to unavailability of vmlinux
    parallels-iso.autogenerated_1: Skipping BTF generation for /usr/lib/parallels-tools/kmods/prl_fs/SharedFolders/Guest/Linux/prl_fs/prl_fs.ko due to unavailability of vmlinux
    parallels-iso.autogenerated_1: Skipping BTF generation for /usr/lib/parallels-tools/kmods/prl_notifier/Installation/lnx/prl_notifier/prl_notifier.ko due to unavailability of vmlinux
    parallels-iso.autogenerated_1: Skipping BTF generation for /usr/lib/parallels-tools/kmods/prl_fs_freeze/Snapshot/Guest/Linux/prl_freeze/prl_fs_freeze.ko due to unavailability of vmlinux
    parallels-iso.autogenerated_1: Start installation of prl_tg kernel module
    parallels-iso.autogenerated_1: Start installation of prl_fs kernel module
    parallels-iso.autogenerated_1: Start installation of prl_fs_freeze kernel module
    parallels-iso.autogenerated_1: Start installation of prl_notifier kernel module
    parallels-iso.autogenerated_1: DKMS modules were added successfully
    parallels-iso.autogenerated_1: DKMS modules for kernel 5.14.0-70.13.1.el9.aarch64 were built successfully
    parallels-iso.autogenerated_1: DKMS modules for kernel 5.14.0-70.13.1.el9.aarch64 were installed successfully
    parallels-iso.autogenerated_1: DKMS modules for kernel 5.14.0-70.30.1.el9.aarch64 were built successfully
    parallels-iso.autogenerated_1: DKMS modules for kernel 5.14.0-70.30.1.el9.aarch64 were installed successfully
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:13: Error: duplicate definition of container_runtime_domtrans(). Original definition on 13.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:40: Error: duplicate definition of container_runtime_run(). Original definition on 40.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:62: Error: duplicate definition of container_runtime_exec(). Original definition on 62.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:81: Error: duplicate definition of container_read_state(). Original definition on 81.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:99: Error: duplicate definition of container_search_lib(). Original definition on 99.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:118: Error: duplicate definition of container_exec_lib(). Original definition on 118.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:137: Error: duplicate definition of container_read_lib_files(). Original definition on 137.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:156: Error: duplicate definition of container_read_share_files(). Original definition on 156.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:177: Error: duplicate definition of container_runtime_read_tmpfs_files(). Original definition on 177.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:198: Error: duplicate definition of container_manage_share_files(). Original definition on 198.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:219: Error: duplicate definition of container_manage_share_dirs(). Original definition on 219.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:239: Error: duplicate definition of container_exec_share_files(). Original definition on 239.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:257: Error: duplicate definition of container_manage_config_files(). Original definition on 257.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:279: Error: duplicate definition of container_manage_lib_files(). Original definition on 279.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:299: Error: duplicate definition of container_manage_files(). Original definition on 299.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:318: Error: duplicate definition of container_manage_dirs(). Original definition on 318.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:336: Error: duplicate definition of container_manage_lib_dirs(). Original definition on 336.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:372: Error: duplicate definition of container_lib_filetrans(). Original definition on 372.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:390: Error: duplicate definition of container_read_pid_files(). Original definition on 390.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:409: Error: duplicate definition of container_systemctl(). Original definition on 409.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:434: Error: duplicate definition of container_rw_sem(). Original definition on 434.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:453: Error: duplicate definition of container_append_file(). Original definition on 453.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:471: Error: duplicate definition of container_use_ptys(). Original definition on 471.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:489: Error: duplicate definition of container_filetrans_named_content(). Original definition on 489.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:563: Error: duplicate definition of container_stream_connect(). Original definition on 563.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:584: Error: duplicate definition of container_spc_stream_connect(). Original definition on 584.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:604: Error: duplicate definition of container_admin(). Original definition on 604.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:654: Error: duplicate definition of container_auth_domtrans(). Original definition on 654.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:673: Error: duplicate definition of container_auth_exec(). Original definition on 673.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:692: Error: duplicate definition of container_auth_stream_connect(). Original definition on 692.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:711: Error: duplicate definition of container_runtime_typebounds(). Original definition on 711.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:730: Error: duplicate definition of container_runtime_entrypoint(). Original definition on 730.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:737: Error: duplicate definition of docker_exec_lib(). Original definition on 737.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:741: Error: duplicate definition of docker_read_share_files(). Original definition on 741.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:745: Error: duplicate definition of docker_exec_share_files(). Original definition on 745.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:749: Error: duplicate definition of docker_manage_lib_files(). Original definition on 749.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:754: Error: duplicate definition of docker_manage_lib_dirs(). Original definition on 754.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:758: Error: duplicate definition of docker_lib_filetrans(). Original definition on 758.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:762: Error: duplicate definition of docker_read_pid_files(). Original definition on 762.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:766: Error: duplicate definition of docker_systemctl(). Original definition on 766.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:770: Error: duplicate definition of docker_use_ptys(). Original definition on 770.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:774: Error: duplicate definition of docker_stream_connect(). Original definition on 774.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:778: Error: duplicate definition of docker_spc_stream_connect(). Original definition on 778.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:792: Error: duplicate definition of container_spc_read_state(). Original definition on 792.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:811: Error: duplicate definition of container_runtime_domain_template(). Original definition on 811.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:849: Error: duplicate definition of container_domain_template(). Original definition on 849.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:877: Error: duplicate definition of container_spc_rw_pipes(). Original definition on 877.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:895: Error: duplicate definition of container_kubelet_domtrans(). Original definition on 895.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:920: Error: duplicate definition of container_kubelet_run(). Original definition on 920.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:940: Error: duplicate definition of container_kubelet_stream_connect(). Original definition on 940.
    parallels-iso.autogenerated_1: Failed to resolve typeattributeset statement at /var/lib/selinux/targeted/tmp/modules/400/prltimesync/cil:35
    parallels-iso.autogenerated_1: Failed to resolve AST
    parallels-iso.autogenerated_1: semodule:  Failed!
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:13: Error: duplicate definition of container_runtime_domtrans(). Original definition on 13.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:40: Error: duplicate definition of container_runtime_run(). Original definition on 40.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:62: Error: duplicate definition of container_runtime_exec(). Original definition on 62.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:81: Error: duplicate definition of container_read_state(). Original definition on 81.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:99: Error: duplicate definition of container_search_lib(). Original definition on 99.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:118: Error: duplicate definition of container_exec_lib(). Original definition on 118.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:137: Error: duplicate definition of container_read_lib_files(). Original definition on 137.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:156: Error: duplicate definition of container_read_share_files(). Original definition on 156.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:177: Error: duplicate definition of container_runtime_read_tmpfs_files(). Original definition on 177.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:198: Error: duplicate definition of container_manage_share_files(). Original definition on 198.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:219: Error: duplicate definition of container_manage_share_dirs(). Original definition on 219.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:239: Error: duplicate definition of container_exec_share_files(). Original definition on 239.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:257: Error: duplicate definition of container_manage_config_files(). Original definition on 257.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:279: Error: duplicate definition of container_manage_lib_files(). Original definition on 279.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:299: Error: duplicate definition of container_manage_files(). Original definition on 299.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:318: Error: duplicate definition of container_manage_dirs(). Original definition on 318.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:336: Error: duplicate definition of container_manage_lib_dirs(). Original definition on 336.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:372: Error: duplicate definition of container_lib_filetrans(). Original definition on 372.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:390: Error: duplicate definition of container_read_pid_files(). Original definition on 390.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:409: Error: duplicate definition of container_systemctl(). Original definition on 409.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:434: Error: duplicate definition of container_rw_sem(). Original definition on 434.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:453: Error: duplicate definition of container_append_file(). Original definition on 453.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:471: Error: duplicate definition of container_use_ptys(). Original definition on 471.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:489: Error: duplicate definition of container_filetrans_named_content(). Original definition on 489.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:563: Error: duplicate definition of container_stream_connect(). Original definition on 563.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:584: Error: duplicate definition of container_spc_stream_connect(). Original definition on 584.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:604: Error: duplicate definition of container_admin(). Original definition on 604.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:654: Error: duplicate definition of container_auth_domtrans(). Original definition on 654.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:673: Error: duplicate definition of container_auth_exec(). Original definition on 673.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:692: Error: duplicate definition of container_auth_stream_connect(). Original definition on 692.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:711: Error: duplicate definition of container_runtime_typebounds(). Original definition on 711.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:730: Error: duplicate definition of container_runtime_entrypoint(). Original definition on 730.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:737: Error: duplicate definition of docker_exec_lib(). Original definition on 737.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:741: Error: duplicate definition of docker_read_share_files(). Original definition on 741.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:745: Error: duplicate definition of docker_exec_share_files(). Original definition on 745.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:749: Error: duplicate definition of docker_manage_lib_files(). Original definition on 749.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:754: Error: duplicate definition of docker_manage_lib_dirs(). Original definition on 754.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:758: Error: duplicate definition of docker_lib_filetrans(). Original definition on 758.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:762: Error: duplicate definition of docker_read_pid_files(). Original definition on 762.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:766: Error: duplicate definition of docker_systemctl(). Original definition on 766.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:770: Error: duplicate definition of docker_use_ptys(). Original definition on 770.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:774: Error: duplicate definition of docker_stream_connect(). Original definition on 774.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:778: Error: duplicate definition of docker_spc_stream_connect(). Original definition on 778.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:792: Error: duplicate definition of container_spc_read_state(). Original definition on 792.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:811: Error: duplicate definition of container_runtime_domain_template(). Original definition on 811.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:849: Error: duplicate definition of container_domain_template(). Original definition on 849.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:877: Error: duplicate definition of container_spc_rw_pipes(). Original definition on 877.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:895: Error: duplicate definition of container_kubelet_domtrans(). Original definition on 895.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:920: Error: duplicate definition of container_kubelet_run(). Original definition on 920.
    parallels-iso.autogenerated_1: /usr/share/selinux/devel/include/services/container.if:940: Error: duplicate definition of container_kubelet_stream_connect(). Original definition on 940.
    parallels-iso.autogenerated_1: Failed to resolve typeattributeset statement at /var/lib/selinux/targeted/tmp/modules/400/prltoolsd/cil:51
    parallels-iso.autogenerated_1: Failed to resolve AST
    parallels-iso.autogenerated_1: semodule:  Failed!
    parallels-iso.autogenerated_1: Parallels Guest Tools were installed successfully!
```

**Comment by paulmaunders** [comment]
Interesting - this shows the same error, but your vagrant box seems to work fine at the end!

**Comment by marcinbojko** [comment]
Even in your case - your output ends with
```
  parallels-iso: Parallels Guest Tools were installed successfully!
```

So the question is - what doesn't work after generation?

**Comment by paulmaunders** [comment]
`vagrant up` hangs when trying to SSH into the box so I think the ssh keys have not been installed correctly, which I assume is related to the parallels tools not fully installing.

```
% vagrant up
Bringing machine 'default' up with 'parallels' provider...
==> default: Registering VM image from the base box 'almalinux-9.0-aarch64'...
==> default: Creating new virtual machine as a linked clone of the box image...
==> default: Unregistering the box VM image...
==> default: Setting the default configuration for VM...
==> default: Setting the name of the VM: test-vagrant-almalinux9_default_1669301893186_26303
==> default: Preparing network interfaces based on configuration...
    default: Adapter 0: shared
==> default: Clearing any previously set network interfaces...
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: :22
    default: SSH username: vagrant
    default: SSH auth method: private key
    default: Warning: Connection refused. Retrying...
    default: Warning: Authentication failure. Retrying...
    default: Warning: Authentication failure. Retrying...
    default: Warning: Authentication failure. Retrying...
    default: Warning: Authentication failure. Retrying...```

**Comment by marcinbojko** [comment]
this doesn't look like a problem with modules, more like the vagrant itself.

**Comment by marcinbojko** [comment]
AFIK keys are being injected the moment machine is being up/

**Comment by paulmaunders** [comment]
Yes it seems to be a problem with the ssh key insertion. 
* [Your Rocky 9 vagrant box](https://app.vagrantup.com/marcinbojko/boxes/rockylinux9_arm64) - key insertion works and shared folders work ('/vagrant/')
* [AlmaLimux 9's aarch64 vagrant box](https://app.vagrantup.com/almalinux/boxes/9.aarch64) - key insertion works, but it fails to mount shared folders
* My own AlmaLinux 9 vagrant box built with Bento - key insertion fails

I'm not sure what you did differently, but yours is the first RHEL 9 vagrant box that has fully worked!

**Comment by marcinbojko** [comment]
I can share vagrantfiles with you

**Comment by marcinbojko** [comment]
https://gist.github.com/marcinbojko/f10ec6b4e778ee3aa0d3fbe680380d43

**Comment by paulmaunders** [comment]
OK so setting config.ssh.password = "vagrant"` in the Vagrantfile does help get further with my AlmaLinux 9 vagrant box, it gets past the SSH step and it also looks like the shared folders also work correctly:

```
paulmaunders@Pauls-MacBook-Air-M2 test-packer-homemade-almalinux9 % vagrant up 
Bringing machine 'default' up with 'parallels' provider...
==> default: Registering VM image from the base box 'almalinux-9.0-aarch64'...
==> default: Creating new virtual machine as a linked clone of the box image...
==> default: Unregistering the box VM image...
==> default: Setting the default configuration for VM...
==> default: Setting the name of the VM: test-packer-homemade-almalinux9_default_1669641236910_91187
==> default: Preparing network interfaces based on configuration...
    default: Adapter 0: shared
==> default: Clearing any previously set network interfaces...
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: :22
    default: SSH username: vagrant
    default: SSH auth method: password
    default: Warning: Connection refused. Retrying...
    default: 
    default: Inserting generated public key within guest...
==> default: Machine booted and ready!
==> default: Checking for Parallels Tools installed on the VM...
==> default: Mounting shared folders...
    default: /vagrant => /Users/paulmaunders/Documents/Code/test-packer-homemade-almalinux9```
    
However, there is still something wrong with the public key insertion because it requires a password to log in, and when I'm in I can see the authorized_keys file is empty.

**Comment by marcinbojko** [comment]
Definitelty Vagramt issue

**Comment by paulmaunders** [comment]
I think I have worked it out. It looks like its something to do with the order the provisioners are run. 

If vagrant.sh is run after update.sh it doesn't work:

```
==> parallels-iso: Provisioning with shell script: /Users/paulmaunders/Documents/Code/test-packer-basic/_common/vagrant.sh
==> parallels-iso: + HOME_DIR=/home/vagrant
==> parallels-iso: + pubkey_url=https://raw.githubusercontent.com/hashicorp/vagrant/master/keys/vagrant.pub
==> parallels-iso: + mkdir -p /home/vagrant/.ssh
==> parallels-iso: + command -v wget
==> parallels-iso: + wget --no-check-certificate https://raw.githubusercontent.com/hashicorp/vagrant/master/keys/vagrant.pub -O /home/vagrant/.ssh/authorized_keys
==> parallels-iso: --2022-11-30 15:42:01--  https://raw.githubusercontent.com/hashicorp/vagrant/master/keys/vagrant.pub
==> parallels-iso: Resolving raw.githubusercontent.com (raw.githubusercontent.com)...
```

If vagrant.sh is run before update.sh then the key insertion works.

```
==> parallels-iso: Provisioning with shell script: /Users/paulmaunders/Documents/Code/test-packer-basic/_common/vagrant.sh
==> parallels-iso: + HOME_DIR=/home/vagrant
==> parallels-iso: + pubkey_url=https://raw.githubusercontent.com/hashicorp/vagrant/master/keys/vagrant.pub
==> parallels-iso: + mkdir -p /home/vagrant/.ssh
==> parallels-iso: + command -v wget
==> parallels-iso: + wget --no-check-certificate https://raw.githubusercontent.com/hashicorp/vagrant/master/keys/vagrant.pub -O /home/vagrant/.ssh/authorized_keys
==> parallels-iso: --2022-11-30 16:32:33--  https://raw.githubusercontent.com/hashicorp/vagrant/master/keys/vagrant.pub
==> parallels-iso: Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
==> parallels-iso: Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
==> parallels-iso: HTTP request sent, awaiting response... 200 OK
==> parallels-iso: Length: 409 [text/plain]
==> parallels-iso: Saving to: ‘/home/vagrant/.ssh/authorized_keys’
==> parallels-iso:
==> parallels-iso:      0K                                                       100% 66.9M=0s
==> parallels-iso:
==> parallels-iso: 2022-11-30 16:32:34 (66.9 MB/s) - ‘/home/vagrant/.ssh/authorized_keys’ saved [409/409]
==> parallels-iso:
==> parallels-iso: + chown -R vagrant /home/vagrant/.ssh
==> parallels-iso: + chmod -R go-rwsx /home/vagrant/.ssh
```


I'm not entirely sure why at the moment. It could be to do with the fact the update.sh is reboots the machine, and perhaps it the network isn't fully working at the key insertion step.

**Comment by paulmaunders** [comment]
I have moved the update.sh out to it's own provision block with a `"pause_after": "30s"` setting and this seems to work! https://github.com/paulmaunders/bento-almalinux9-arm-only/commit/e8b3d8e9d30bade7b3739c42866ee4b145abdca0

**Comment by marcinbojko** [comment]
@paulmaunders I am deadly confused.
I do not use update.sh or vagrant.sh.
Is it still connected with my repository and my scripts?

**Comment by paulmaunders** [comment]
Hi @marcinbojko sorry to confuse you!

This is not actually an issue with your scripts or box.  It's a problem with the ssh key insertion in the Bento packer templates and Almalinux Cloud Images. 

Your vagrant box works fine, in fact it was the only working image I could find, so I created this question to try to find out why yours worked, and the others didn't. 

I have found the problem with those build systems. It's to do with the network not being ready after update reboot. I have fixed it with the pause_after so we can close this now.

Thanks for your help in debugging this!
